packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,305 @@
+}
+
+interface RouteInfo extends ConfigRoute {
+  name: string;
+  segments: string[];
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory - The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routeMap = getRouteMap(appDirectory, routePaths, prefix);
+
+  let uniqueRoutes = new Map<string, string>();
+
+  function defineNestedRoutes(
+    defineRoute: DefineRouteFunction,
+    parentId?: string
+  ): void {
+    let childRoutes = Array.from(routeMap.values()).filter(
+      (routeInfo) => routeInfo.parentId === parentId
+    );
+    let parentRoute = parentId ? routeMap.get(parentId) : undefined;
+    let parentRoutePath = parentRoute?.path ?? "/";
+    for (let childRoute of childRoutes) {
+      let routePath = childRoute?.path?.slice(parentRoutePath.length) ?? "";
+      // remove leading slash
+      if (routePath.startsWith("/")) {
+        routePath = routePath.slice(1);
+      }
+      let index = childRoute.index;
+      let fullPath = childRoute.path;
+
+      // add option to check for unique route ids
+      // this is copied from remix default convention
+      // but it is currently breaking some flat routes
+      // so until we can figure out a better way to do this
+      // make it optional to unblock users
+      let uniqueRouteId = (fullPath || "") + (index ? "?index" : "");
+      if (uniqueRouteId) {
+        let conflict = uniqueRoutes.get(uniqueRouteId);
+        if (conflict) {
+          throw new Error(
+            `Path ${JSON.stringify(fullPath)} defined by route ${JSON.stringify(
+              childRoute.id
+            )} conflicts with route ${JSON.stringify(conflict)}`
+          );
+        } else {
+          uniqueRoutes.set(uniqueRouteId, childRoute.id);
+        }
+      }
+      if (index) {
+        let invalidChildRoutes = Object.values(routeMap).filter(
+          (routeInfo) => routeInfo.parentId === childRoute.id
+        );
+
+        if (invalidChildRoutes.length > 0) {
+          throw new Error(
+            `Child routes are not allowed in index routes. Please remove child routes of ${childRoute.id}`
+          );
+        }
+
+        defineRoute(routePath, routeMap.get(childRoute.id!)!.file, {
+          index: true,
+        });
+      } else {
+        defineRoute(routePath, routeMap.get(childRoute.id!)!.file, () => {
+          defineNestedRoutes(defineRoute, childRoute.id);
+        });
+      }
+    }
+  }
+
+  let routes = defineRoutes(defineNestedRoutes);
+
+  Object.values(routes).forEach((route) => {
+    if (route.parentId === undefined) {
+      route.parentId = "root";
+    }
+  });
```

Comment:
I'm surprised that setting `parentId` to `"root"` for missing parents isn't already done by `defineRoutes`.

Is that something that should be added to `defineRoutes` so that this doesn't have to be patched here? Or is there a reason to want to keep this logic outside of `defineRoutes`?